bitnami/kafka/templates/kafka-provisioning.yaml

```diff
@@ -0,0 +1,91 @@
+{{- if .Values.provisioning.enabled }}
+kind: Job
+apiVersion: batch/v1
+metadata:
+  name: {{ include "kafka.provisioning.fullname" . }}
+  labels: {{- include "common.labels.standard" . | nindent 4 }}
+    app.kubernetes.io/component: kafka
+    {{- if .Values.commonLabels }}
+    {{- include "common.tplvalues.render" ( dict "value" .Values.commonLabels "context" $ ) | nindent 4 }}
+    {{- end }}
+  annotations:
+    helm.sh/hook: post-install,post-upgrade
+    helm.sh/hook-delete-policy: before-hook-creation
+    {{- if .Values.commonAnnotations }}
+    {{- include "common.tplvalues.render" ( dict "value" .Values.commonAnnotations "context" $ ) | nindent 4 }}
+    {{- end }}
+spec:
+  template:
+    metadata:
+      labels: {{- include "common.labels.standard" . | nindent 8 }}
+        app.kubernetes.io/component: kafka
+        {{- if .Values.podLabels }}
+        {{- include "common.tplvalues.render" (dict "value" .Values.podLabels "context" $) | nindent 8 }}
+        {{- end }}
+      annotations:
+    spec:
+{{- include "kafka.imagePullSecrets" . | indent 6 }}
+      restartPolicy: OnFailure
+      terminationGracePeriodSeconds: 0
+      containers:
+        - name: kafka-provisioning
+          image: {{ include "kafka.provisioning.image" . }}
+          imagePullPolicy: {{ .Values.provisioning.image.pullPolicy | quote }}
+          command:
+            - /bin/bash
+            - -c
+            - >-
+            {{- $bootstrapServer := printf "%s:%d" (include "kafka.fullname" .) (.Values.service.port | int64) }}
+            {{- range $topic := .Values.provisioning.topics }}
+              /opt/bitnami/kafka/bin/kafka-topics.sh \
+                --create \
+                --if-not-exists \
+                --bootstrap-server {{ $bootstrapServer }} \
+                --replication-factor {{ $topic.replicationFactor }} \
+                --partitions {{ $topic.partitions }} \
+                {{- range $name, $value := $topic.config }}
+                --config {{ $name }}={{ $value }} \
+                {{- end }}
+                --topic {{ $topic.name }};
+            {{- end }}
```

Comment:
Can you explain what topic injection failure means? Does `set -e` already fixes this problem?

Reply:
Sure can, I gave this a run the other night and I had a spelling mistake in my values.yaml which was throwing an error in the pod logs, but the pod was exiting with zero due to the echo at the end exiting with a zero. So the restart on failure never triggered and rescheduled the job. The thinking I have is we can reliably fail until the server is up and we can inject the topics or until the helm deployment timeouts in which case helm should fail because the job isn't completed.

Reply:
Hmm, if you want non zero status codes on job error - set -e will propogate it

This job is a helm hook, if it fails - release will fail

Reply:
Hi @sergeyshaykhullin,

We should add `&&` to `--topic {{ $topic.name }} && ` instead of `;` as there are instances where:  the job installs one topic -> gets an error for the other one -> job restarts -> reruns both creation commands successfully -> the logs only shows `Created 2nd topic` as the first one was already created.

Reply:
Hmm, i am not really good at bash, but... Does && fix something?

First operation completed && second failed
Restart
First operation without stdout && second operation completed

I will test does && fix it. I thought that --if-not-exists if... exists prints nothing and exit

Reply:
It seems it won't solve my corner case, though at least it would restart the job if its the firts topic the one that fails. That way, only when the first topic is installed will the job continue.

Reply:
To restart job on first (N) failure i used set -e

Reply:
For now topics provisioning works sequentially

Do you mean you want to provision topics is parallel and if any failed - try again?

Reply:
That would be ideal, as the topic's creation will always be reflected on the logs in its entirety.

Reply:
No, when you create topics in parallel, you can also lose logs. Even more

Start creating 10 topics
9 created, 1 failed
Restart job, lose 9 topic creation logs
1 topic created

Reply:
You are right, I completely misunderstood that code. I was thinking of a way to have all the created topics on the job's log. What do you think about adding something like:

```
              echo "Provisioning succeeded.";
              /opt/bitnami/kafka/bin/kafka-topics.sh \
                --bootstrap-server {{ $bootstrapServer }} \
                --list
```

Reply:
Can you please provide a use-case for listing all topics? Does kafka-topics.sh supports somehow listing newly created topics only?

Reply:
It would be just for readability as it would avoid confusion. Although it would list all the existing topics and not only the newly created, it may be a compromise to not having the stdout for those topics created before the pod's last restart.

Reply:
I think current approach is good. If you creating topic - you will get message, if no message for topic you want - means it is already created

For readability i think before each kafka-topics run i can add something like 'echo Ensure {topic} created'

If topic exists - next ensure message logged
If topic created - kafka-topics will print message before 'ensure' message

Reply:
Yes, I think that could be a good compromise. Thank you for your work!